src/llms/openai.ts

```diff
@@ -388,3 +390,153 @@ export const enabled = async (): Promise<OpenAIHealthDetails> => {
     { configured: details.openAI, ok: details.openAI } :
     details.openAI;
 }
+
+export enum StreamStatus {
+  IDLE = 'idle',
+  GENERATING = 'generating',
+  COMPLETED = 'completed',
+}
+
+export const TIMEOUT = 10000;
+
+
+export type OpenAIStreamState = {
+  setMessages: React.Dispatch<React.SetStateAction<Message[]>>;
+  reply: string;
+  streamStatus: StreamStatus;
+  error: Error | undefined;
+  value:
+    | {
+        enabled: boolean | undefined;
+        stream?: undefined;
+      }
+    | {
+        enabled: boolean | undefined;
+        stream: Subscription;
+      }
+    | undefined;
+}
+
+/**
+ * A custom React hook for managing an OpenAI stream that communicates with the provided model.
+ *
+ * @param {string} [model=DEFAULT_OAI_MODEL] - The OpenAI model to use for communication.
+ * @param {number} [temperature=1] - The temperature value for text generation (default is 1).
+ * @param {function} [notifyError] - A callback function for handling errors.
+ *
+ * @returns {OpenAIStreamState} - An object containing the state of the OpenAI stream.
+ * @property {function} setMessages - A function to update the list of messages in the stream.
+ * @property {string} reply - The most recent reply received from the OpenAI stream.
+ * @property {StreamStatus} streamStatus - The status of the stream ("idle", "generating" or "completed").
+ * @property {Error|undefined} error - An error object if an error occurs, or undefined if no error.
+ * @property {object|undefined} value - The current value of the stream.
+ * @property {boolean|undefined} value.enabled - Indicates whether the stream is enabled (true or false).
+ * @property {Subscription|undefined} value.stream - The stream subscription object if the stream is active, or undefined if not.
+ */
```

Comment:
Not sure how easy this will be but could you add a super minimal example to the JSDoc? It can be similar to the examples in the other exposed functions.

Reply:
I don't think this is feasible because there is so much handling to do, but once we have a working example elsewhere perhaps we can link it?